vumi_http_proxy/http_proxy.py

```diff
@@ -1,47 +1,109 @@
+from twisted.python import log
 from twisted.web import http, proxy
 from twisted.internet import reactor
 from twisted.internet.endpoints import serverFromString
+from twisted.names import client
+from twisted.web.client import Agent, readBody
+from urlparse import urlparse, urlunparse
+from twisted.internet.defer import inlineCallbacks, succeed
 
-# blacklist of disallowed domains (change this to ips later, and move)
+# blacklist of disallowed domains (move to proxy_blacklist.py)
 
-DEFAULT_BLACKLIST = ["facebook.com", "twitter.com", "zombo.com"]
-
-#  Set up proxy
+DEFAULT_BLACKLIST = ["69.16.230.117"]
 
 
 class ProxyFactory(http.HTTPFactory):
 
-    def __init__(self, blacklist):
+    def __init__(self, blacklist, resolver, http_client):
         http.HTTPFactory.__init__(self)
         self.blacklist = blacklist
+        self.resolver = resolver
+        self.http_client = http_client
 
     def buildProtocol(self, addr):
-        return Proxy(self.blacklist)
+        return Proxy(self.blacklist, self.resolver, self.http_client)
 
 
-# Check request
 class CheckProxyRequest(proxy.ProxyRequest):
     def process(self):
         host, _, port = self.getAllHeaders()['host'].partition(':')
-        if host in self.channel.blacklist:
+        d = self.channel.resolver.getHostByName(host)
+        d.addCallback(self.setIP, host)
+        d.addErrback(self.handleError)
+        return d
+
+    def handleError(self, failure):
+        log.err(failure)
+        self.setResponseCode(400)
+        self.write("<html>Denied</html>")
+        self.finish()
+        return
+
+    def setIP(self, ip_addr, host):
+        if not ip_addr:
+            self.setResponseCode(400)
+            self.write("<html>ERROR: No IP adresses found for name %r" %
+                       host + " </html>")
+            self.finish()
+            return
+        if ip_addr in DEFAULT_BLACKLIST:
             self.setResponseCode(400)
             self.write("<html>Denied</html>")
             self.finish()
             return
-        # else allow access to site
-        return proxy.ProxyRequest.process(self)
+        else:
```

Comment:
Since we `return` at the end of the `if` blocks above, we could leave out the `else` and unindent the rest of the function. Not vital, but some people prefer that because its less nesting and makes what is happening a bit clearer to them (some people probably includes me :).